### ISSUE TITLE
Now CLI 17.0.3 seems to ignore rewrites in dev

### ISSUE THREAD

**Issue by wiesson** [vercel/vercel#3777]
I have noticed that `now dev` (17.0.3) ignores the rewrites from the `now.json` config file. I have downgraded to 16.7.3 and it works as expected. I'm using the rewrites to proxy requests through our domain

```
{
  "version": 2,
  "builds": [
    {
      "src": "next.config.js",
      "use": "@now/next"
    }
  ],
  "rewrites": [
    {
      "source": "/api/v1/graphql",
      "destination": "https://route.to/our/backend/api/endpoint"
    }
  ]
}
```

Maybe this was deprecated or there is another breaking changes, but it's quite hard to see what is actually breaking change, the 17.0.0 release just says:

> Now CLI 17.0.0 is here! This includes a few breaking changes including Project Linking.

A few means, there are more breaking changes? :)

**Comment by mirshko** [comment]
Just having this issue too.

Im also not able to get env vars in my project. This is a serious blocker for me.

They only work in production. Now Dev is ignoring it all.

**Comment by wiesson** [comment]
@mirshko - regarding env vars -> https://github.com/zeit/now/issues/3758

**Comment by glennreyes** [comment]
Redirects also seem to be ignored with 17.0.3.

**Comment by willemliu** [comment]
For me it seems to not work with `now dev` and production as well. It simply feels like the rewrites, redirects, headers are completely ignored.
I asked this on the Zeit Spectrum as well. https://spectrum.chat/zeit/now/now-dev-v17-0-3-now-json-routes-give-404~d8c8ce9b-c655-4a2f-b593-968f2f090886

I need the rewrites part to work because I make use of next-offline which generates a service-worker in a particular place. So I need to rewrite `/service-worker.js` to something like `/_next/static/service-worker.js`.

Perhaps this bug is related to this one: https://github.com/zeit/now/issues/3539

**Comment by wiesson** [comment]
I haven't had the time to debug the issue and I simply switched back to `16.7.3` to get things done. But I might look into it soon.

**Comment by DarthRukian** [comment]
Still not working in 17.0.4 rewrites and headers (to set custom content-type) are not working in `now dev` .
Any chance to be fixed soon ?

Thanks

**Comment by HP4k1h5** [comment]
Ive been having similar issues since very recently. [post](https://spectrum.chat/zeit/now/router-external-target-error~af8633d5-8c6f-4891-9445-249b94c9c569)
Im getting a 502 bad gateway response. The response to external api requests from deployed instances is 
```
An internal error occurred with ZEIT Now.
ROUTER_EXTERNAL_TARGET_ERROR
```

**Comment by willemliu** [comment]
I've received an response from Zeit support:

> So I consulted with the team and indeed there was a major change from version 16 and 17 of the Now CLI, which may have broken your `now dev` experience. In version 17, we don't build the Next.js outputs and instead proxy the next dev server for a more stable dev experience. This is why it from now on you should be specifying your rewrites/redirects in next.config.js instead of now.json(for Next.js projects). Can you please upgrade to the latest canary version of  Next.js (9.2.3-canary.22) and see if it fixes your `service-worker.js` rewrite?  As I mentioned, it's still experimental and it will hopefully be in a stable release soon.

Implementation of the experimental rewrites feature in `next.config.js`.

```
module.exports = {
    experimental: {
      async rewrites() {
        return [
         { source: '/podcasts', destination: '/podcasts-overview' }
        ]
      }
    }
}
```

**Comment by raybotha** [comment]
Not working for me on experimental, 17 or 16. :man_shrugging:

**Comment by DarthRukian** [comment]
The solution worked for me :) You need next in 9.2.3-canary.22 and now in 17.0.4

**Comment by leo** [comment]
Thank you for sharing!

If you're running into a problem with `rewrites`, `redirects`, `headers`, `cleanUrls` or `trailingSlash`, we recommend using `routes` instead – which is [documented in detail here](https://zeit.co/docs/configuration#project/routes).

Should `routes` not work, please make sure to let us know so we can take a look immediately. Additionally, using the `rewrites` and `redirects` offered by Next.js is a great idea (we always recommend using framework features over platform features if your framework supports it).

We're making sure that, this week, the documentation and Now CLI are updated to reflect the suggested way of configuring routing. At the moment, as you know, we're offering two varieties of routing properties, which can be confusing.

**Comment by rfgamaral** [comment]
I was under the impression that `routes` were being deprecated??

**Comment by guidolodetti** [comment]
~@rfgamaral yes, routes are deprecated but rewrites are not yet working well 😅~ seems not (read below)

**Comment by styfle** [comment]
No, `routes` are not deprecated.

**Comment by willemliu** [comment]
> No, `routes` are not deprecated.

Indeed! It was however recommended to use `rewrites`, `headers` and `redirects` according to their documentation until the 6th of March. See commit link below to the documentation. And when using those properties it is no longer possible to use `routes`. It seems that the new properties are not stable enough. I may have remembered wrong. But I seem to recall seeing a warning about `routes` being deprecated. Maybe it was in the `Now-CLI` `v17.0.1` when `routes` are still in the configuration.

6 March - Removed recommendation: https://github.com/zeit/docs/commit/641a6234c57be78c4efebe4ee12aabe37bcd2928

**Comment by brunosalgueiro** [comment]
### I think **"routes, rewrites and redirects"** are a very important feature from now.
Everyone who updates **NOW** and **NOW CLI** to 17+ is breaking their app's locally. Is there any place telling us about removing this feature?
Will NOW CLI continue to use the routes in now.json or has it been decided to be deprecated?
Why there is no backward compatibility?

I would like to ask help about this, we need more information about these changes (if you know more, please share with us).

I would like to thank you very much @willemliu for share your "experimental" solution for the problem, your post is helping everyone.
"Experimental" Solution to Next.js: https://spectrum.chat/zeit/now/now-dev-v17-0-3-now-json-routes-give-404~d8c8ce9b-c655-4a2f-b593-968f2f090886

More questions about: https://github.com/zeit/now/issues/3789#issuecomment-588963045

**Comment by theKashey** [comment]
I've updated `now.sh` a few days ago and run into this issue.
Actually I did not understand what is happening, and why it's always displaying `index.html`, instead of going through API layer.
So:
- I've renamed `public/index.html` to `public/main.html`, but `now` still somehow discovered the "right" file and displayed me
- I've replaced `rewrites` by `routes` and __removed__ `{ "handle": "filesystem" }` - and it still works! It's better say - nothing changed.

In other words - `routes` do not work as well. 
> Now CLI  17.1.1 dev (beta)

**Comment by wiesson** [comment]
There is an open RFC which moves the rewrites / routes to nextjs -> https://github.com/zeit/next.js/issues/9081

**Comment by theKashey** [comment]
But what if you are not using `next.js`? 
The `now` itself is not bound to the javascript, and could handle different platforms as well.

**Comment by henderea** [comment]
I agree, it would be a very bad idea to remove routes and related configs from `now` just because they're supported in `next.js`.  I personally have never used `next.js`, and it isn't really something I plan to use, so it is important to me to not lose config options just because one specific framework supports them.  I often use static deployments, but those deployments might use `react` and `react-router`, so I would still need to be able to configure routes, rewrites, and other configs in `now`.  Honestly, I pretty much exclusively use static deployments with `now`, so losing support for using things like `react-router` and `window.history.push` in a static deployment would hurt my ability to properly use `now`.

**Comment by janzheng** [comment]
hmm are there any migration guides out there? seems like the new now breaks `now dev` too (using Vue/nuxtjs, not next)

**Comment by theKashey** [comment]
Spent another hour trying to migrate to v19. But nor rewrites, nor routes are working for me. Access to functions under `/api` not working as well - I am always getting just my `main.html`

**Comment by GxDesign** [comment]
> I've received an response from Zeit support:
> 
> > So I consulted with the team and indeed there was a major change from version 16 and 17 of the Now CLI, which may have broken your `now dev` experience. In version 17, we don't build the Next.js outputs and instead proxy the next dev server for a more stable dev experience. This is why it from now on you should be specifying your rewrites/redirects in next.config.js instead of now.json(for Next.js projects). Can you please upgrade to the latest canary version of  Next.js (9.2.3-canary.22) and see if it fixes your `service-worker.js` rewrite?  As I mentioned, it's still experimental and it will hopefully be in a stable release soon.
> 
> Implementation of the experimental rewrites feature in `next.config.js`.
> 
> ```
> module.exports = {
>     experimental: {
>       async rewrites() {
>         return [
>          { source: '/podcasts', destination: '/podcasts-overview' }
>         ]
>       }
>     }
> }
> ```

Seems like this would break NextJS multi-zone support

**Comment by gunar** [comment]
In my experience `rewrite` was not being observed by Vercel on `vercel dev`. However, if I touched `vercel.json`, the dev script would restart the process automatically (as it should) and the rules were then indeed followed. That being said, in preview/production it still wouldn't work.

I've made the following replacements (as according to the [`routes` docs](https://vercel.com/docs/configuration#project/routes))
- `rewrites` => `routes`
- `source` => `src`
- `destination` => `dest`

It works now. My problem is solved.

**Comment by ekawatani** [comment]
When I do `vercel dev` with `cleanUrls` set to true, it redirects `/about.html` to `/about`, but it doesn't do `/about` to `/about.html`. I'm using next.js and I have the HTML file under `public` directory. Something is still broken here.

I'm using vercel version 21.0.1.

**Comment by andymckay** [comment]
It's been quite a while since this issue and we are on version 28.x of the CLI. I suggest at this point if there's any more issues to open a new issue or support ticket here https://vercel.com/help#issues and we can take a look.
